drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +45,494 @@ extern "C" {
+        .read_word(caboose_magic_offset as u32)
+        .map_err(|_| RawCabooseError::MissingCaboose)?;
+
+    if caboose_magic == CABOOSE_MAGIC {
+        let caboose_range = ((caboose_magic_offset as u32) + U32_SIZE)
+            ..(caboose_size_offset as u32);
+        Ok(caboose_range)
+    } else {
+        Err(RawCabooseError::MissingCaboose)
+    }
 }
 
-/// Return the flash storage address range and flash execution address range.
-/// These are only different for the staged stage0 image.
-pub fn image_range(
-    component: RotComponent,
-    slot: SlotId,
-) -> (Range<u32>, Range<u32>) {
-    unsafe {
-        match (component, slot) {
-            (RotComponent::Hubris, SlotId::A) => (
-                __IMAGE_A_BASE.as_ptr() as u32..__IMAGE_A_END.as_ptr() as u32,
-                __IMAGE_A_BASE.as_ptr() as u32..__IMAGE_A_END.as_ptr() as u32,
-            ),
-            (RotComponent::Hubris, SlotId::B) => (
-                __IMAGE_B_BASE.as_ptr() as u32..__IMAGE_B_END.as_ptr() as u32,
-                __IMAGE_B_BASE.as_ptr() as u32..__IMAGE_B_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::A) => (
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::B) => (
-                __IMAGE_STAGE0NEXT_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0NEXT_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
+/// Accessor keeps the implementation details of ImageAccess private
+enum Accessor<'a> {
+    // Flash driver, flash device range
+    Flash {
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        span: FlashRange,
+    },
+    Ram {
+        buffer: &'a [u8],
+        span: FlashRange,
+    },
+    // Hybrid is used for later implementation of rollback protection.
+    // The buffer is used in place of the beginning of the flash range.
+    _Hybrid {
+        buffer: &'a [u8],
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        span: FlashRange,
+    },
+}
+
+impl Accessor<'_> {
+    fn exec(&self) -> &Range<u32> {
+        match self {
+            Accessor::Flash { flash: _, span }
+            | Accessor::Ram { buffer: _, span }
+            | Accessor::_Hybrid {
+                buffer: _,
+                flash: _,
+                span,
+            } => &span.exec,
+        }
+    }
+}
+
+/// The update_server needs to deal with images that are:
+///   - Entirely in flash
+///   - Header block in RAM with the remainder unavailable
+///   - Header block in RAM with the remainder in flash.
+///   - Entire image in RAM (cached Stage0)
+/// Calls to methods use image relative addresses.
+/// ImageAccess::Flash{_, span} gives the physical flash addresses
```

Comment:
What is `ImageAccess::Flash{_, span}`? I can't make sense of it.

Reply:
Old comment fragments. Edited to make sense.